lesson18/script.js

```diff
@@ -0,0 +1,186 @@
+const slider = document.getElementById("js-slider");
+
+const createAttributedElements = ({tag,attrObj,str}) =>{
+    const element = document.createElement(tag);
+    Object.keys(attrObj).forEach((attribute) => {
+        element.setAttribute(attribute, attrObj[attribute]);
+    });
+    if(str !== undefined) element.textContent = str;
+    return element;
+}
+
+const createErrorMessage = (error) => {
+    const errMsg = createAttributedElements({tag:"p",attrObj:{class:"error-msg"},str:error});
+    
+    slider.appendChild(errMsg);
+}
+
+const showLoading = () => {
+    const loadingPlace = createAttributedElements({tag:"div",attrObj:{id:"js-loading",class:"loading"}});
+    const loadingImage = createAttributedElements({tag:"img",attrObj:{src:"img/loading-circle.gif",alt:"ローディング"}});
+    
+    slider.appendChild(loadingPlace).appendChild(loadingImage);
+}
+
+const removeLoading = () => document.getElementById("js-loading").remove();
+
+const fetchData = async (URL) => {
+    const response = await fetch(URL);
+    if(!response.ok){
+        console.error(`${response.status}:${response.statusText}`);
+        createErrorMessage('現在、サーバーの通信が壊れています');
+    }
+    const json = await response.json();
+    const data = json.data;
+    return data;
+}
+
+const fetchSliderData = (ms) => new Promise(resolve => setTimeout(() => resolve(fetchData("https://myjson.dit.upm.es/api/bins/2bzx")),ms));
+
+const init = async () => {
+    showLoading();
+    try{
+        const sliderData = await fetchSliderData(3000);
+        if (!sliderData) {
+            return;
+        }
+        renderSliderElements(sliderData);
+        autoSwitchSlider(sliderData);
+    }catch(error){
+        console.error(error);
+        createErrorMessage(error);
+    }finally{
+        removeLoading();
+    }
+}
+
+const renderSliderElements = (sliderData) => {
+    renderSliderItems(sliderData);
+    getPageNumber(sliderData);
+    renderButtons(sliderData);
+    renderDotPagination(sliderData);
+    isActiveFirstItem();
+}
+
+let autoPlay;
+const autoSwitchSlider = (sliderData) =>{
+    autoPlay = setInterval(() =>{
+        currentIndex ++;
+        if(currentIndex >= sliderData.length){
+            currentIndex = 0;
+        }
+        switchSlider(sliderData,currentIndex);
+    }, 3000);
+}
+
+const resetAutoPlay = (sliderData) => {
+    clearInterval(autoPlay);
+    autoSwitchSlider(sliderData);
+}
+
+const renderSliderItems = (sliderData) => {
+    const fragment = document.createDocumentFragment();
+    for(const image of sliderData){
+        const sliderItem = createAttributedElements({tag:"div",attrObj:{class:"slider__item"}});
+        const slideImage = createAttributedElements({tag:"img",attrObj:{class:"slide-image",src:image.img,alt:image.alt}});
```

Comment:
Q.
I felt uncomfortable with the HTML structure of the slider.
I felt that `<ul><li><img>...` was the standard.
Please let me know if I'm wrong :)!!

Reply:
Absolutely.
I think the shape of `<ul> <li> <img>` is correct for slider.
Thank you for teaching me.
I fixed.
https://github.com/nakagawamai/js-lesson/pull/20/commits/274be37fdd11e48a95869a573456ea78f96e90b7